### PR TITLE
Few issues solved - separator/DRY/list(...)

### PR DESCRIPTION
**Pull request by Polkas**
I think it is enough for this PR, new ideas I will push in another one.

- [x] separator https://github.com/daroczig/logger/issues/95 
After many tests I think that the new utils function `catch_base_log` is the best way to get log header length (nchar). It might return wrong number of chars if log header has different length under different environments. 
Another idea which was not implemented is to add `width` argument to logger -> log_level. Thus control message/output length at the creation level.
I provide informative roxygen2 docs when log_separator is not perfectly supported.
- [x] list(...) https://github.com/daroczig/logger/issues/92 For list(..) I transfer part of the code.
- [x] DRY rule for log_* family functions - a new template function `log_base`
- [x] ifelse -> if else - where input is singular. `ifelse` is recommended for vectors where as for single logical we should use `if else`

### CODE REVIEW

File: R/appenders.R [hunk 1]
```diff
@@ -89,7 +89,7 @@ appender_file <- function(file, append = TRUE, max_lines = Inf, max_bytes = Inf,
                 n_lines <- tryCatch(
                     suppressWarnings(R.utils::countLines(file)),
                     error = function(e) 0)
-                n_bytes <- ifelse(file.exists(file), file.info(file)$size, 0)
+                n_bytes <- if (file.exists(file)) file.info(file)$size else 0
```
Comment (daroczig) [inline R/appenders.R:92]: what's the advantage of calling `if/else` (without the curly braces) compared to `ifelse`?

Comment (Polkas) [reply R/appenders.R:92]: It is more secure and readable. 
https://twitter.com/hadleywickham/status/928983687605059585?lang=bn

Comment (Polkas) [reply R/appenders.R:92]: It is more secure and readable.
Here a Hadley Wickham opinion who is in favour of such step.
https://twitter.com/hadleywickham/status/928983687605059585

Comment (tdeenes) [reply R/appenders.R:92]: Strongly agree with @Polkas . Note also this edge case (not relevant for `file.exists(file)`, since it can not return `NA`, but still):
```
> ifelse(NA, TRUE, FALSE)
[1] NA
> `if`(NA, TRUE, FALSE)
Error in if (NA) TRUE else FALSE : missing value where TRUE/FALSE needed
```
Note also that for simple conditionals, I tend to use backticked `if` to keep the functional style.

Comment (daroczig) [reply R/appenders.R:92]: Hm, got it, thanks for the pointers. Although makes sense from a perspective, but I have an aversion to using `if` without the curly braces, as have been hurt by how the parser handles it (single VS multiple statements), e.g. with `covr` false positives or missed `else` branches on multiline expressions.

Although I've not run into this issue recently, habits die hard :) 

And the R manual of `if` still suggests always keeping the braces .. which makes it ugly for single-line statements.

Anyway, I will keep thinking about it, and looking at the rest of the PR in the meanwhile -- thanks again!!

File: R/utils.R [hunk 2]
```diff
@@ -54,3 +54,21 @@ deparse_to_one_line <- function(x) {
...
          perl = TRUE)
 }
+
+#' Catch the log header
+#' @return string
+#' @keywords internal
+#' @param level see \code{\link{log_levels}}
+#' @param namespace string
+catch_base_log <- function(level, namespace) {
+    orginal_appender <- log_appender()
```
Comment (daroczig) [inline R/utils.R:64]: don't we need to pass `namespace` here and below as well to the `log_appender` calls?

File: R/utils.R [hunk 3]
```diff
@@ -54,3 +54,21 @@ deparse_to_one_line <- function(x) {
...
+    log_appender(appender_console)
+    # catch error, warning or message
+    res <- capture.output(log_level(level = level, namespace = namespace), type = "message")
+    if (length(res) == 0) {
+        # catch output
+        res <- capture.output(log_level(level = level, namespace = namespace), type = "output")
+    }
```
Comment (daroczig) [inline R/utils.R:71]: do we need to check both `stderr` and `stdout`? setting `appender_console` (which is an alias to `appender_stderr`) should (in theory) write to the former, and nothing to the latter.

File: R/appenders.R (hunk 1 above)
Comment (Polkas) [reply R/appenders.R:92]: I understand your doubts, so what about `ifelse(isTRUE(cond), ..., ...)`.

File: R/utils.R (hunk 3 above)
Comment (Polkas) [reply R/utils.R:71]: I added the `log_appender(appender_console)` call as I got double prints in some tests. Preciously for helpers testing where there is used `log_appender(appender_stdout)`.  `catch_base_log` should be neutral for the rest of the session, that is why I am doing stderr redirection (2>).

File: R/utils.R (hunk 2 above)
Comment (Polkas) [reply R/utils.R:64]: Connected with https://github.com/daroczig/logger/pull/96#discussion_r760845468. This is only a dummy setup to send everything to stderr.

File: R/utils.R (hunk 3 above)
Comment (Polkas) [reply R/utils.R:71]: hmmm maybe sth like:
```
orginal_appender <- log_appender()
log_appender(logger::appender_stderr)
# catch error, warning or message
res <- capture.output(log_level(level = level, namespace = namespace), type = "message")
log_appender(logger::appender_stdout)
res <- capture.output(log_level(level = level, namespace = namespace), type = "output")
log_appender(orginal_appender)
res
```

Comment (daroczig) [reply R/utils.R:71]: double prints should not happen, except if namespaces and/or indexes are messed up. I think passing `namespace` in the other related comment, and systematically everywhere in this function might resolve it?

Comment (Polkas) [reply R/utils.R:71]: I think this is ok as log_appender is additional print and not influence the base one.
```
> logger::log_appender(logger::appender_stderr)
> logger::log_error("aa")
ERROR [2021-12-02 09:28:17] aa
> logger::log_info("aa")
INFO [2021-12-02 09:28:20] aa
```

File: R/logger.R [hunk 4]
```diff
@@ -304,34 +268,35 @@ log_namespaces <- function() {
 log_level <- function(level, ..., namespace = NA_character_,
                       .logcall = sys.call(), .topcall = sys.call(-1), .topenv = parent.frame()) {
 
+    log_arg <- list(...)
```
Comment (daroczig) [inline R/logger.R:271]: awesome, thanks!

File: R/logger.R [hunk 5]
```diff
@@ -67,11 +67,50 @@ fallback_namespace <- function(namespace) {
...
 }
 
+#' Base Logging Function
+#' @param fun_name string a full name of log function
+#' @param arg see \code{\link{log_levels}}
+#' @param namespace logger namespace
+#' @param index index of the logger within the namespace
+#' @return currently set or return log function property
+#' @keywords internal
+log_base <- function(fun_name, arg, namespace, index) {
```
Comment (daroczig) [inline R/logger.R:77]: I'm thinking about renaming to something like `log_config_setter` -- thoughts?

File: R/logger.R [hunk 6]
```diff
@@ -67,11 +67,50 @@ fallback_namespace <- function(namespace) {
...
+    }
+
+    configs[[min(index, length(config) + 1)]] <- config
+    assign(namespace, configs, envir = namespaces)
+}
 
 #' Get or set log level threshold
-#' @param level see \code{\link{log_levels}}
+#' @param level see \code{\link{log_levels}}, default NULL
 #' @param namespace logger namespace
```
Comment (daroczig) [inline R/logger.R:114]: `namespace` and `index` can `inheritParams` from the base setter

Comment (Polkas) [reply R/logger.R:114]: reverted

File: R/logger.R (hunk 5 above)
Comment (Polkas) [reply R/logger.R:77]: `log_config_setter` sounds nice, applied

File: R/utils.R (hunk 3 above)
Comment (Polkas) [reply R/utils.R:71]: You are right namespaces were a clue. More than that I have to use fallback_namespace.

```r
catch_base_log <- function(level, namespace) {
    namespace <- fallback_namespace(namespace)
    orginal_appender <- log_appender(namespace = namespace)
    log_appender(appender_console, namespace = namespace)
    # catch error, warning or message
    res <- capture.output(log_level(level = level, namespace = namespace), type = "message")
    log_appender(orginal_appender, namespace = namespace)
    res
}
```

1. getting namespace and if not exists then return global.
2. getting current namespace appender, to be able to return to it later.
3. temporally setting appender_console (stderr) inside the namespace and catch (run) capture.output(log_layout(level, namespace = namespace), type = "message").
4. Go back to original appender and apply it to the namespace
5. return catched header (from log_layout)

We get a log_layout header and do not influence the environment

Comment (Polkas) [reply R/utils.R]: Please read on github as I added small edit.

Comment (Polkas) [reply R/utils.R]: In the last commit I added testthat tests for the `catch_base_log` function.

Comment (Polkas) [reply R/utils.R]: I start to consider `.topcall` so the function is even more secure when the layout depends on where the function is called. I added `.topcall` argument to log_separator too. More tests are added for this scenerios.

File: R/utils.R (hunk 2 above)
Comment (Polkas) [reply R/utils.R]: applied

**Comment by daroczig** [comment]
sorry for the delay with this - got distracted by #98 and the weekend .. I will pick it up again in a few days

File: R/appenders.R (hunk 1 above)
Comment (daroczig) [reply R/appenders.R:92]: Sorry, let's not wrap `file.exists` in `isTRUE` :zipper_mouth_face: 

I don't think there's any benefit in that extra check, still makes the code more complex and a bit slower.

So far, the backtick + `if` combo is the most appealing to me .. but still feels like black magic, so probably we need something easier to grasp -- which might be just the ugly multi-line `if` conditionals?

Anyway, this is more of a language syntax question, and I feel like we just miss a nice R implementation of the ternary operator, which is unlikely that we will resolve in this thread :) 

So I might be open to updating the above to avoid `if` conditionals altogether with something like:

```r
na.omit(c(file.info(file)$size, 0))[1]
```

But let me rather check the rest of the PR instead at last :)

File: R/helpers.R [hunk 7]
```diff
@@ -84,10 +84,27 @@ log_eval <- function(expr, level = TRACE, multiline = FALSE) {
...
-log_separator <- function(level = INFO, namespace = NA_character_, separator = '=', width = 80) {
+log_separator <- function(
+  level = INFO, 
+  namespace = NA_character_,
+  separator = '=',
+  width = 80,
+  .topcall = sys.call()
+  ) {
+
+    stopifnot(length(separator) == 1 && nchar(separator) == 1)
```
Comment (daroczig) [inline R/helpers.R:102]:
```suggestion
    stopifnot(length(separator) == 1, nchar(separator) == 1)
```

**Comment by daroczig** [comment]
This is great, thank you very much again! Also thanks for the ping in the other PR on the Travis -> GH actions update :+1:

I am ready to merge this, but may I ask the below changes before that?

- Can you please revert the `ifelse` updates? I don't see a clear winner at the time being, so let's delay that discussion to a future PR.
- Can you please revert the code style changes in the old code? E.g. breaking long roxygen lines. And update the newly added lines to conform with the rest of the codebase: single quotes, 4 spaces.

No problem if not, I can totally do that, just let me know.

**Comment by Polkas** [comment]
I reverted ifelse, " quote and roxygen2 long lines split. Moreover I go through new code and update it to use 4 spaces. 

Have a great new year:)

File: tests/testthat/test-utils.R [hunk 8]
```diff
@@ -24,5 +24,21 @@ test_that('validate_log_level', {
...
 
+test_that('catch_base_log', {
+    expect_true(nchar(logger:::catch_base_log(ERROR, NA_character_)) == 28)
+    expect_true(nchar(logger:::catch_base_log(INFO, NA_character_)) == 27)
+    layout_original <- log_layout()
+    log_layout(layout_blank)
+    expect_true(nchar(logger:::catch_base_log(INFO, NA_character_)) == 0)
+    log_layout(layout_original)
+    layout_original <- log_layout(namespace = 'TEMP')
+    logger <- layout_glue_generator(format = '{node}/{pid}/{namespace}/{fn} {time} {level}: {msg}')
```
Comment (daroczig) [inline tests/testthat/test-utils.R:35]: I suspect the PR builder checks (tests) are failing due to this .. `nod` and `pid` can have variable with, maybe test without those?

Comment (daroczig) [reply tests/testthat/test-utils.R:35]: see e.g. https://github.com/daroczig/logger/runs/4681051142?check_suite_focus=true

Comment (Polkas) [reply tests/testthat/test-utils.R]: I change the computer and now it works on both of them. You are right node and pid have to be removed from tests as are machine specific. Good finding, you save a lot of my time:) Of course now it looks to be obvious.

Comment (Polkas) [reply tests/testthat/test-utils.R]: image

Comment (Polkas) [reply tests/testthat/test-utils.R]: ah looks like default `time` has different format on old Linux machine. I will remove it from layout and we should be at home. Rest of machines are ok.

**Comment by Polkas** [comment]
Unfortunately we find out some edge case. 
Luckily I have a pythonanywhere account, with a Linux and the old R version (3.6.3).

Looks like we find some edge case.

image

**Comment by daroczig** [comment]
I am pretty sure that is due to `testtat` messing with the scopes, see e.g. https://github.com/daroczig/logger/blob/master/tests/testthat/test-CRANSKIP-logger-namespaces.R#L18

One workaround, just like above, is making the call happen outside of `testthat` .. or just remove the `fn` from the check as well :zipper_mouth_face:

**Comment by Polkas** [comment]
Finally I removed one line (one test), as I think we should still test how`.topcall` influence the call length.
If you prefer other solution, of course I am leaving it to you.)

**Review by daroczig — APPROVED** [review]
Thanks a lot again for this, and sorry for the back-and-forths :sweat_smile: This was a great idea and excellent contribution :bow:
